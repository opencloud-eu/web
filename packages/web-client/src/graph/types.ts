import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
