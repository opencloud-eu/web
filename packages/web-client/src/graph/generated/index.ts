/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

