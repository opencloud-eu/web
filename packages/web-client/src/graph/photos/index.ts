export * from './photo'
export * from './types'
