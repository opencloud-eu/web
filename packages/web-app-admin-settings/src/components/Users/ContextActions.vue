<template>
  <div>
    <context-action-menu :menu-sections="menuSections" :action-options="{ resources: items }" />
  </div>
</template>

<script lang="ts">
import {
  useUserActionsEdit,
  useUserActionsDelete,
  useUserActionsEditQuota
} from '../../composables/actions/users'
import { computed, defineComponent, PropType, unref } from 'vue'
import { ContextActionMenu } from '@opencloud-eu/web-pkg'
import { User } from '@opencloud-eu/web-client/graph/generated'
import { useActionsShowDetails } from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'ContextActions',
  components: { ContextActionMenu },
  props: {
    items: {
      type: Array as PropType<User[]>,
      required: true
    }
  },
  setup(props) {
    const filterParams = computed(() => ({ resources: props.items }))

    const { actions: showDetailsActions } = useActionsShowDetails()
    const { actions: editQuotaActions } = useUserActionsEditQuota()
    const { actions: userEditActions } = useUserActionsEdit()
    const { actions: userDeleteActions } = useUserActionsDelete()

    const menuItemsPrimaryActions = computed(() =>
      [...unref(userEditActions), ...unref(userDeleteActions)].filter((item) =>
        item.isVisible(unref(filterParams))
      )
    )
    const menuItemsSecondaryActions = computed(() =>
      [...unref(editQuotaActions)].filter((item) => item.isVisible(unref(filterParams)))
    )

    const menuItemsSidebar = computed(() =>
      [...unref(showDetailsActions)].filter((item) => item.isVisible(unref(filterParams)))
    )

    const menuSections = computed(() => {
      const sections = []

      if (unref(menuItemsPrimaryActions).length) {
        sections.push({
          name: 'primaryActions',
          items: unref(menuItemsPrimaryActions)
        })
      }
      if (unref(menuItemsSecondaryActions).length) {
        sections.push({
          name: 'secondaryActions',
          items: unref(menuItemsSecondaryActions)
        })
      }
      if (unref(menuItemsSidebar).length) {
        sections.push({
          name: 'sidebar',
          items: unref(menuItemsSidebar)
        })
      }
      return sections
    })

    return {
      menuSections
    }
  }
})
</script>
