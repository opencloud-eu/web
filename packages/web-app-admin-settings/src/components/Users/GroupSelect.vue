<template>
  <div id="user-group-select-form">
    <oc-select
      :model-value="selectedOptions"
      class="oc-mb-s"
      :multiple="true"
      :options="groupOptions"
      option-label="displayName"
      :label="$gettext('Groups')"
      :fix-message-line="true"
      v-bind="$attrs"
      :required-mark="requiredMark"
      @update:model-value="onUpdate"
    >
      <template #selected-option="{ displayName, id }">
        <span class="oc-flex oc-flex-center">
          <avatar-image
            class="oc-flex oc-align-self-center oc-mr-s"
            :width="16.8"
            :userid="id"
            :user-name="displayName"
          />
          <span>{{ displayName }}</span>
        </span>
      </template>
      <template #option="{ displayName, id }">
        <div class="oc-flex">
          <span class="oc-flex oc-flex-center">
            <avatar-image
              class="oc-flex oc-align-self-center oc-mr-s"
              :width="16.8"
              :userid="id"
              :user-name="displayName"
            />
            <span>{{ displayName }}</span>
          </span>
        </div>
      </template>
    </oc-select>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, PropType, ref, unref, watch } from 'vue'
import { Group } from '@opencloud-eu/web-client/graph/generated'

export default defineComponent({
  name: 'GroupSelect',
  props: {
    selectedGroups: {
      type: Array as PropType<Group[]>,
      required: true
    },
    groupOptions: {
      type: Array as PropType<Group[]>,
      required: true
    },
    requiredMark: {
      type: Boolean,
      default: false,
      required: false
    }
  },
  emits: ['selectedOptionChange'],
  setup(props, { emit }) {
    const selectedOptions = ref()
    const onUpdate = (group: Group) => {
      selectedOptions.value = group
      emit('selectedOptionChange', unref(selectedOptions))
    }

    const currentGroups = computed(() => props.selectedGroups)
    watch(
      currentGroups,
      () => {
        selectedOptions.value = props.selectedGroups
          .map((g) => ({
            ...g,
            readonly: g.groupTypes?.includes('ReadOnly')
          }))
          .sort((a: any, b: any) => b.readonly - a.readonly)
      },
      { immediate: true }
    )

    return { selectedOptions, onUpdate }
  }
})
</script>
