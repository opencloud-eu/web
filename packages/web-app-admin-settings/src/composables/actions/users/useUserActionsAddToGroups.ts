import { computed, Ref, unref } from 'vue'
import { useGettext } from 'vue3-gettext'
import { UserAction, useModals, useCapabilityStore, UserActionOptions } from '@opencloud-eu/web-pkg'
import { Group } from '@opencloud-eu/web-client/graph/generated'
import AddToGroupsModal from '../../../components/Users/AddToGroupsModal.vue'

export const useUserActionsAddToGroups = ({ groups }: { groups: Ref<Group[]> }) => {
  const { dispatchModal } = useModals()
  const { $gettext, $ngettext } = useGettext()
  const capabilityStore = useCapabilityStore()

  const handler = ({ resources }: UserActionOptions) => {
    dispatchModal({
      title: $ngettext(
        'Add user "%{user}" to groups',
        'Add %{userCount} users to groups ',
        resources.length,
        {
          user: resources[0].displayName,
          userCount: resources.length.toString()
        }
      ),
      customComponent: AddToGroupsModal,
      customComponentAttrs: () => ({
        users: resources,
        groups: unref(groups)
      })
    })
  }

  const actions = computed((): UserAction[] => [
    {
      name: 'add-to-groups',
      icon: 'add',
      class: 'oc-users-actions-add-to-groups-trigger',
      label: () => $gettext('Add to groups'),
      isVisible: ({ resources }) => {
        if (capabilityStore.graphUsersReadOnlyAttributes.includes('user.memberOf')) {
          return false
        }

        return resources.length > 0
      },
      handler
    }
  ])

  return {
    actions
  }
}
