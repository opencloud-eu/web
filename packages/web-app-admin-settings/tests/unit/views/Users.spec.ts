import Users from '../../../src/views/Users.vue'
import { ItemFilter, OptionsConfig, UserAction, useAppDefaults } from '@opencloud-eu/web-pkg'
import { mock, mockDeep } from 'vitest-mock-extended'
import {
  defaultComponentMocks,
  defaultPlugins,
  mount,
  shallowMount,
  useAppDefaultsMock
} from '@opencloud-eu/web-test-helpers'
import { ClientService, queryItemAsString } from '@opencloud-eu/web-pkg'
import { Group, User } from '@opencloud-eu/web-client/graph/generated'
import { useUserActionsCreateUser } from '../../../src/composables/actions/users/useUserActionsCreateUser'
import { ref } from 'vue'

vi.mock('@opencloud-eu/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  queryItemAsString: vi.fn(),
  useAppDefaults: vi.fn()
}))
vi.mock('../../../src/composables/actions/users/useUserActionsCreateUser')
vi.mocked(useAppDefaults).mockImplementation(() => useAppDefaultsMock())

const getDefaultUser = (): User => {
  return {
    id: '1',
    displayName: 'Admin',
    givenName: 'Admin',
    surname: 'Admin',
    memberOf: [],
    mail: 'admin@example.org',
    drive: {
      id: '1',
      name: 'admin',
      quota: { remaining: 5000000000, state: 'normal', total: 5000000000, used: 0 }
    },
    appRoleAssignments: [
      {
        appRoleId: '1',
        id: '1',
        principalId: '1',
        principalType: 'User',
        resourceDisplayName: 'OpenCloud',
        resourceId: 'some-graph-app-id'
      }
    ]
  } as User
}

const getDefaultApplications = () => {
  return [
    {
      appRoles: [
        {
          displayName: 'Admin',
          id: '1'
        },
        {
          displayName: 'Guest',
          id: '2'
        },
        {
          displayName: 'Space Admin',
          id: '3'
        },
        {
          displayName: 'User',
          id: '4'
        }
      ],
      displayName: 'OpenCloud',
      id: 'some-graph-app-id'
    }
  ]
}

const getClientService = () => {
  const clientService = mockDeep<ClientService>()
  clientService.graphAuthenticated.users.listUsers.mockResolvedValue([mock<User>(getDefaultUser())])
  clientService.graphAuthenticated.users.getUser.mockResolvedValue(mock<User>(getDefaultUser()))
  clientService.graphAuthenticated.users.editUser.mockResolvedValue(mock<User>(getDefaultUser()))
  clientService.graphAuthenticated.groups.listGroups.mockResolvedValue([mock<Group>()])
  clientService.graphAuthenticated.applications.listApplications.mockResolvedValue(
    getDefaultApplications()
  )
  return clientService
}

const selectors = {
  itemFilterGroupsStub: 'item-filter-stub[filtername="groups"]',
  itemFilterRolesStub: 'item-filter-stub[filtername="roles"]',
  createUserButton: '#create-user-btn'
}

describe('Users view', () => {
  describe('list view', () => {
    it('renders list initially', async () => {
      const { wrapper } = getMountedWrapper({ mountType: mount, users: [getDefaultUser()] })
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('renders initially warning if filters are mandatory', async () => {
      const { wrapper } = getMountedWrapper({
        mountType: mount,
        options: { userListRequiresFilter: true }
      })
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('create user button', () => {
    it('should be displayed if action enabled', () => {
      const { wrapper } = getMountedWrapper({
        mountType: mount,
        createUserActionEnabled: true
      })
      const createUserButton = wrapper.find(selectors.createUserButton)
      expect(createUserButton.exists()).toBeTruthy()
    })
    it('should not be displayed if action disabled', () => {
      const { wrapper } = getMountedWrapper({
        mountType: mount,
        createUserActionEnabled: false
      })
      const createUserButton = wrapper.find(selectors.createUserButton)
      expect(createUserButton.exists()).toBeFalsy()
    })
  })

  describe('computed method "sideBarAvailablePanels"', () => {
    it('should contain EditPanel when one user is selected', () => {
      const { wrapper } = getMountedWrapper()
      expect(
        wrapper.vm.sideBarAvailablePanels
          .find(({ name }) => name === 'EditPanel')
          .isVisible({ items: [{ id: '1' } as User] })
      ).toBeTruthy()
    })
    it('should contain DetailsPanel no user is selected', () => {
      const { wrapper } = getMountedWrapper()
      expect(
        wrapper.vm.sideBarAvailablePanels
          .find(({ name }) => name === 'DetailsPanel')
          .isVisible({ items: [] })
      ).toBeTruthy()
    })
    it('should not contain EditPanel when multiple users are selected', () => {
      const { wrapper } = getMountedWrapper()
      expect(
        wrapper.vm.sideBarAvailablePanels
          .find(({ name }) => name === 'EditPanel')
          .isVisible({ items: [{ id: '1' }, { id: '2' }] as User[] })
      ).toBeFalsy()
    })
  })

  describe('batch actions', () => {
    it('do not display when no user selected', async () => {
      const { wrapper } = getMountedWrapper({ mountType: mount })
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.find('batch-actions-stub').exists()).toBeFalsy()
    })
    it('display when one user selected', async () => {
      const { wrapper } = getMountedWrapper({
        mountType: mount,
        selectedUsers: [{ id: '1' } as User]
      })
      await wrapper.vm.loadResourcesTask.last
      await wrapper.vm.$nextTick()
      expect(wrapper.find('batch-actions-stub').exists()).toBeTruthy()
    })
    it('display when more than one users selected', async () => {
      const { wrapper } = getMountedWrapper({
        mountType: mount,
        selectedUsers: [{ id: '1' }, { id: '2' }] as User[]
      })
      await wrapper.vm.loadResourcesTask.last
      await wrapper.vm.$nextTick()
      expect(wrapper.find('batch-actions-stub').exists()).toBeTruthy()
    })
  })

  describe('filter', () => {
    describe('groups', () => {
      it('does filter users by groups when the "selectionChange"-event is triggered', async () => {
        const clientService = getClientService()
        const { wrapper } = getMountedWrapper({ mountType: mount, clientService })
        await wrapper.vm.loadResourcesTask.last
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledTimes(1)
        wrapper
          .findComponent<typeof ItemFilter>(selectors.itemFilterGroupsStub)
          .vm.$emit('selectionChange', [{ id: '1' }])
        await wrapper.vm.$nextTick()
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledTimes(2)
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenNthCalledWith(
          2,
          {
            orderBy: ['displayName'],
            filter: "(memberOf/any(m:m/id eq '1'))",
            expand: ['appRoleAssignments']
          },
          expect.anything()
        )
      })
      it('does filter initially if group ids are given via query param', async () => {
        const groupIdsQueryParam = '1+2'
        const clientService = getClientService()
        const { wrapper } = getMountedWrapper({
          mountType: mount,
          clientService,
          groupFilterQuery: groupIdsQueryParam
        })
        await wrapper.vm.loadResourcesTask.last
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledWith(
          {
            orderBy: ['displayName'],
            filter: "(memberOf/any(m:m/id eq '1') or memberOf/any(m:m/id eq '2'))",
            expand: ['appRoleAssignments']
          },
          expect.anything()
        )
      })
    })
    describe('roles', () => {
      it('does filter users by roles when the "selectionChange"-event is triggered', async () => {
        const clientService = getClientService()
        const { wrapper } = getMountedWrapper({ mountType: mount, clientService })
        await wrapper.vm.loadResourcesTask.last
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledTimes(1)
        wrapper
          .findComponent<typeof ItemFilter>(selectors.itemFilterRolesStub)
          .vm.$emit('selectionChange', [{ id: '1' }])
        await wrapper.vm.$nextTick()
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledTimes(2)
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenNthCalledWith(
          2,
          {
            orderBy: ['displayName'],
            filter: "(appRoleAssignments/any(m:m/appRoleId eq '1'))",
            expand: ['appRoleAssignments']
          },
          expect.anything()
        )
      })
      it('does filter initially if role ids are given via query param', async () => {
        const roleIdsQueryParam = '1+2'
        const clientService = getClientService()
        const { wrapper } = getMountedWrapper({
          mountType: mount,
          clientService,
          roleFilterQuery: roleIdsQueryParam
        })
        await wrapper.vm.loadResourcesTask.last
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledWith(
          {
            orderBy: ['displayName'],
            filter:
              "(appRoleAssignments/any(m:m/appRoleId eq '1') or appRoleAssignments/any(m:m/appRoleId eq '2'))",
            expand: ['appRoleAssignments']
          },
          expect.anything()
        )
      })
    })
    describe('displayName', () => {
      it('does filter initially if displayName is given via query param', async () => {
        const displayNameFilterQueryParam = 'Albert'
        const clientService = getClientService()
        const { wrapper } = getMountedWrapper({
          mountType: mount,
          clientService,
          displayNameFilterQuery: displayNameFilterQueryParam
        })
        await wrapper.vm.loadResourcesTask.last
        expect(clientService.graphAuthenticated.users.listUsers).toHaveBeenCalledWith(
          {
            orderBy: ['displayName'],
            filter: "contains(displayName,'Albert')",
            expand: ['appRoleAssignments']
          },
          expect.anything()
        )
      })
    })
  })
})

function getMountedWrapper({
  mountType = shallowMount,
  clientService = getClientService(),
  displayNameFilterQuery = null,
  groupFilterQuery = null,
  roleFilterQuery = null,
  options = {},
  createUserActionEnabled = true,
  users = [],
  selectedUsers = []
}: {
  mountType?: typeof shallowMount | typeof mount
  clientService?: ReturnType<typeof mockDeep<ClientService>>
  displayNameFilterQuery?: string
  groupFilterQuery?: string
  roleFilterQuery?: string
  options?: OptionsConfig
  createUserActionEnabled?: boolean
  users?: User[]
  selectedUsers?: User[]
} = {}) {
  vi.mocked(queryItemAsString).mockImplementationOnce(() => displayNameFilterQuery)
  vi.mocked(queryItemAsString).mockImplementationOnce(() => groupFilterQuery)
  vi.mocked(queryItemAsString).mockImplementationOnce(() => roleFilterQuery)
  vi.mocked(queryItemAsString).mockImplementationOnce(() => displayNameFilterQuery)
  vi.mocked(useUserActionsCreateUser).mockReturnValue(
    mock<ReturnType<typeof useUserActionsCreateUser>>({
      actions: ref([mock<UserAction>({ isVisible: () => createUserActionEnabled })])
    })
  )

  const mocks = {
    ...defaultComponentMocks(),
    $clientService: clientService
  }

  const user = { id: '1' } as User

  return {
    mocks,
    wrapper: mountType(Users, {
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              userState: { user },
              configState: { options },
              userSettingsStore: { users, selectedUsers }
            }
          })
        ],
        mocks,
        provide: mocks,
        stubs: {
          AppLoadingSpinner: true,
          ViewOptions: true,
          OcBreadcrumb: true,
          NoContentMessage: true,
          ItemFilter: true,
          BatchActions: true,
          OcButton: true
        }
      }
    })
  }
}
