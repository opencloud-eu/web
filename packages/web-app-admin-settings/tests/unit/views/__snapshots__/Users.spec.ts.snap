// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Users view > list view > renders initially warning if filters are mandatory 1`] = `
"<div data-v-32efd9c6="">
  <main data-v-32efd9c6="" class="oc-flex oc-height-1-1 app-content oc-width-1-1">
    <div class="admin-settings-wrapper oc-flex oc-width-1-1 oc-width-expand oc-height-1-1 oc-flex-wrap">
      <div id="admin-settings-view-wrapper" class="oc-width-expand oc-width-1-1 oc-height-1-1 oc-flex-wrap">
        <div id="admin-settings-app-bar" class="oc-app-bar oc-py-s admin-settings-app-bar-sticky">
          <div class="admin-settings-app-bar-controls oc-flex oc-flex-between oc-flex-middle">
            <oc-breadcrumb-stub items="[object Object],[object Object]" contextmenupadding="medium" id="admin-settings-breadcrumb" maxwidth="-1" showcontextactions="false" truncationoffset="2" variation="default"></oc-breadcrumb-stub>
            <portal-target name="app.runtime.mobile.nav"></portal-target>
            <div class="oc-flex">
              <view-options-stub perpagestorageprefix="admin-settings" hashiddenfiles="false" hasfileextensions="false" haspagination="true" paginationoptions="20,50,100,250" perpagequeryname="items-per-page" perpagedefault="50" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
            </div>
          </div>
          <div class="admin-settings-app-bar-actions oc-flex oc-flex-middle oc-mt-xs">
            <div data-v-32efd9c6="">
              <oc-button-stub data-v-32efd9c6="" appearance="filled" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="create-user-btn" class="oc-mr-s"></oc-button-stub>
            </div>
            <!--v-if-->
          </div>
        </div>
        <div data-v-32efd9c6="" id="user-list" class="">
          <div class="user-filters oc-flex oc-flex-between oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m">
            <div data-v-32efd9c6="" class="oc-flex oc-flex-middle">
              <div data-v-32efd9c6="" class="oc-mr-m oc-flex oc-flex-middle"><span data-v-32efd9c6="" class="oc-icon oc-icon-m oc-mr-xs"><!----></span> <span data-v-32efd9c6="">Filter:</span></div>
              <item-filter-stub data-v-32efd9c6="" filterlabel="Groups" items="undefined" filtername="groups" optionfilterlabel="Filter groups" showoptionfilter="true" allowmultiple="true" idattribute="id" displaynameattribute="displayName" filterableattributes="displayName" closeonclick="false" class="oc-mr-s"></item-filter-stub>
              <item-filter-stub data-v-32efd9c6="" filterlabel="Roles" items="[object Object],[object Object],[object Object],[object Object]" filtername="roles" optionfilterlabel="Filter roles" showoptionfilter="true" allowmultiple="true" idattribute="id" displaynameattribute="displayName" filterableattributes="displayName" closeonclick="false"></item-filter-stub>
            </div>
            <div data-v-32efd9c6="" class="oc-flex oc-flex-middle">
              <div data-v-32efd9c6="" class=""><label class="oc-label" for="users-filter">Search</label>
                <div class="oc-position-relative">
                  <!--v-if--> <input id="users-filter" modelmodifiers="[object Object]" autocomplete="off" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text">
                  <!--v-if-->
                </div>
                <!--v-if-->
                <!--v-if-->
              </div>
              <oc-button-stub data-v-32efd9c6="" appearance="raw" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="users-filter-confirm" class="oc-ml-xs oc-p-xs"></oc-button-stub>
            </div>
          </div>
          <div>
            <no-content-message-stub data-v-32efd9c6="" icon="error-warning" iconfilltype="fill"></no-content-message-stub>
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </main>
</div>"
`;

exports[`Users view > list view > renders list initially 1`] = `
"<div data-v-32efd9c6="">
  <main data-v-32efd9c6="" class="oc-flex oc-height-1-1 app-content oc-width-1-1">
    <div class="admin-settings-wrapper oc-flex oc-width-1-1 oc-width-expand oc-height-1-1 oc-flex-wrap">
      <div id="admin-settings-view-wrapper" class="oc-width-expand oc-width-1-1 oc-height-1-1 oc-flex-wrap">
        <div id="admin-settings-app-bar" class="oc-app-bar oc-py-s admin-settings-app-bar-sticky">
          <div class="admin-settings-app-bar-controls oc-flex oc-flex-between oc-flex-middle">
            <oc-breadcrumb-stub items="[object Object],[object Object]" contextmenupadding="medium" id="admin-settings-breadcrumb" maxwidth="-1" showcontextactions="false" truncationoffset="2" variation="default"></oc-breadcrumb-stub>
            <portal-target name="app.runtime.mobile.nav"></portal-target>
            <div class="oc-flex">
              <view-options-stub perpagestorageprefix="admin-settings" hashiddenfiles="false" hasfileextensions="false" haspagination="true" paginationoptions="20,50,100,250" perpagequeryname="items-per-page" perpagedefault="50" viewmodedefault="resource-table" viewmodes=""></view-options-stub>
            </div>
          </div>
          <div class="admin-settings-app-bar-actions oc-flex oc-flex-middle oc-mt-xs">
            <div data-v-32efd9c6="">
              <oc-button-stub data-v-32efd9c6="" appearance="filled" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="create-user-btn" class="oc-mr-s"></oc-button-stub>
            </div>
            <!--v-if-->
          </div>
        </div>
        <div data-v-32efd9c6="" id="user-list" class="">
          <div class="user-filters oc-flex oc-flex-between oc-flex-wrap oc-flex-bottom oc-mx-m oc-mb-m">
            <div data-v-32efd9c6="" class="oc-flex oc-flex-middle">
              <div data-v-32efd9c6="" class="oc-mr-m oc-flex oc-flex-middle"><span data-v-32efd9c6="" class="oc-icon oc-icon-m oc-mr-xs"><!----></span> <span data-v-32efd9c6="">Filter:</span></div>
              <item-filter-stub data-v-32efd9c6="" filterlabel="Groups" items="undefined" filtername="groups" optionfilterlabel="Filter groups" showoptionfilter="true" allowmultiple="true" idattribute="id" displaynameattribute="displayName" filterableattributes="displayName" closeonclick="false" class="oc-mr-s"></item-filter-stub>
              <item-filter-stub data-v-32efd9c6="" filterlabel="Roles" items="[object Object],[object Object],[object Object],[object Object]" filtername="roles" optionfilterlabel="Filter roles" showoptionfilter="true" allowmultiple="true" idattribute="id" displaynameattribute="displayName" filterableattributes="displayName" closeonclick="false"></item-filter-stub>
            </div>
            <div data-v-32efd9c6="" class="oc-flex oc-flex-middle">
              <div data-v-32efd9c6="" class=""><label class="oc-label" for="users-filter">Search</label>
                <div class="oc-position-relative">
                  <!--v-if--> <input id="users-filter" modelmodifiers="[object Object]" autocomplete="off" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text">
                  <!--v-if-->
                </div>
                <!--v-if-->
                <!--v-if-->
              </div>
              <oc-button-stub data-v-32efd9c6="" appearance="raw" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="users-filter-confirm" class="oc-ml-xs oc-p-xs"></oc-button-stub>
            </div>
          </div>
          <div>
            <table class="oc-table oc-table-hover oc-table-sticky has-item-context-menu users-table">
              <thead class="oc-thead">
                <tr class="oc-table-header-row">
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-select oc-pl-m" style="top: 0px;">
                    <div><span class="oc-table-thead-content"><span><input id="oc-checkbox-1" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" aria-label="Select all users"> <!--v-if--></span></span></div>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-th oc-table-header-cell oc-table-header-cell-avatar" style="top: 0px;">
                    <div><span class="oc-table-thead-content header-text"></span></div>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-onPremisesSamAccountName" style="top: 0px;" aria-sort="ascending">
                    <oc-button-stub arialabel="Sort by onPremisesSamAccountName" appearance="raw" colorrole="secondary" disabled="false" gapsize="small" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="true" class="oc-button-sort oc-width-1-1"></oc-button-stub>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-displayName" style="top: 0px;" aria-sort="none">
                    <oc-button-stub arialabel="Sort by displayName" appearance="raw" colorrole="secondary" disabled="false" gapsize="small" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="true" class="oc-button-sort oc-width-1-1"></oc-button-stub>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-mail" style="top: 0px;" aria-sort="none">
                    <oc-button-stub arialabel="Sort by mail" appearance="raw" colorrole="secondary" disabled="false" gapsize="small" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="true" class="oc-button-sort oc-width-1-1"></oc-button-stub>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-role" style="top: 0px;" aria-sort="none">
                    <oc-button-stub arialabel="Sort by role" appearance="raw" colorrole="secondary" disabled="false" gapsize="small" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="true" class="oc-button-sort oc-width-1-1"></oc-button-stub>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-accountEnabled" style="top: 0px;" aria-sort="none">
                    <oc-button-stub arialabel="Sort by accountEnabled" appearance="raw" colorrole="secondary" disabled="false" gapsize="small" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="true" class="oc-button-sort oc-width-1-1"></oc-button-stub>
                  </th>
                  <th class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-th oc-table-header-cell oc-table-header-cell-actions oc-pr-m" style="top: 0px;">
                    <div><span class="oc-table-thead-content header-text">Actions</span></div>
                  </th>
                </tr>
              </thead>
              <tbody class="has-item-context-menu">
                <tr class="oc-tbody-tr oc-tbody-tr-1" data-item-id="1" draggable="false">
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-select oc-pl-m"><span><input id="oc-checkbox-2" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" aria-label="Select Admin" value="[object Object]"> <!--v-if--></span></td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-shrink oc-td oc-table-data-cell oc-table-data-cell-avatar">
                    <avatar-image width="32" userid="1" user-name="Admin"></avatar-image>
                  </td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-onPremisesSamAccountName"></td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-displayName mark-element">Admin</td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-mail">admin@example.org</td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-role">Admin</td>
                  <td class="oc-table-cell oc-table-cell-align-left oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-accountEnabled"><span class="oc-flex oc-flex-middle"><span class="oc-icon oc-icon-m oc-mr-s"><!----></span><span>Allowed</span></span></td>
                  <td class="oc-table-cell oc-table-cell-align-right oc-table-cell-align-middle oc-table-cell-width-auto oc-td oc-table-data-cell oc-table-data-cell-actions oc-pr-m">
                    <oc-button-stub arialabel="Show details" appearance="raw" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" class="oc-ml-xs quick-action-button oc-p-xs users-table-btn-details"></oc-button-stub>
                    <oc-button-stub arialabel="Edit" appearance="raw" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" class="oc-ml-xs quick-action-button oc-p-xs users-table-btn-edit"></oc-button-stub>
                    <oc-button-stub arialabel="Show context menu" appearance="raw" colorrole="secondary" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="context-menu-trigger-1" class="quick-action-button oc-ml-xs oc-p-xs users-table-btn-action-dropdown"></oc-button-stub>
                  </td>
                </tr>
              </tbody>
              <tfoot class="oc-table-footer">
                <tr class="oc-table-footer-row">
                  <td colspan="8" class="oc-table-footer-cell">
                    <!-- @slot Footer of the table -->
                    <!--v-if-->
                    <div class="oc-text-center oc-width-1-1 oc-my-s">
                      <p class="oc-text-muted">1 users in total</p>
                    </div>
                  </td>
                </tr>
              </tfoot>
            </table>
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </main>
</div>"
`;
